proposals/stack-switching/Explainer.md

```diff
@@ -0,0 +1,876 @@
+`t1*` to a suspended continuation of type `$ct1`, yielding a modified
+suspended continuation of type `$ct2`. The `cont.bind` instruction
+also consumes its continuation argument, and yields a new continuation
+that can be supplied to `resume`,`resume_throw`, `switch` or
+`cont.bind`.
+
+### Continuation lifetime
+
+#### Producing continuations
+
+There are four different ways in which continuations may be produced
+(`cont.new,suspend,cont.bind,switch`). A fresh continuation object
+is allocated with `cont.new` and the current continuation is reused
+with `suspend`, `cont.bind`, and `switch`.
+
+The `cont.bind` instruction is similar to the `func.bind` instruction
+that was initially part of the function references proposal. However,
+whereas the latter necessitates the allocation of a new closure, as
+continuations are single-shot no allocation is necessary: all
+allocation happens when the original continuation is created by
+preallocating one slot for each continuation argument.
+
+#### Consuming continuations
+
+There are four different ways in which suspended continuations are
+consumed (`resume,resume_throw,switch,cont.bind`). A suspended
+continuation may be resumed with a particular handler with `resume`;
+aborted with `resume_throw`; directly switched to via `switch`; or
+partially applied with `cont.bind`.
+
+In order to ensure that continuations are one-shot, `resume`,
+`resume_throw`, `switch`, and `cont.bind` destructively modify the
+suspended continuation such that any subsequent use of the same
+suspended continuation will result in a trap.
+
+## Design considerations
+
+In this section we discuss some key design considerations.
+
+### Asymmetric switching
+
+Resuming a suspended continuation establishes a parent-child
+relationship which aligns with the caller-callee relationship for
+standard function calls meaning that no special plumbing is needed in
+order to compose the non-local control features we define with
+built-in non-local control features such as traps, exceptions, and
+embedder integration.
+
+### Symmetric switching
+
+Direct switching to a suspended peer continuation is semantically
+equivalent to suspending the current continuation with a special
+switch tag whose payload is the suspended peer continuation in the
+context of a handler which resumes the peer continuation. However,
+direct switching can (and should) be optimised to avoid the need to
+switch control to the handler before switching control to the peer.
+
+### Partial application
+
+Partial application can be important in practice due to the block and
+type structure of Wasm, as in order to return a continuation from a
+block all branches within the block must agree on the type of
+continuation. Using `cont.bind`, a producer can ensure that the
+branches within a block each produce a continuation with the same type.
+
+### One-shot continuations
+
+Continuations in the current proposal are single-shot (aka linear),
```

Comment:
Do multi-shot continuations put a significantly larger burden on the implementation?

Is the current proposal suitable to be amended/complemented by a future proposal adding multi-shot continuations?

Reply:
Whether it is difficult to implement depends on the particular semantics for multi-shot continuations and the architectural constraints of the implementing engine.

I do not think there is anything in this proposal that prohibits the future addition of an instruction for simulating multi-shot continuations via cloning, e.g. `cont.clone : [(ref null $ct)] -> [(ref $ct) (ref $ct)]`.

Reply:
Thanks for the response.

Yeah, so like this proposal generalizes the concept of exceptions to non-local control flow, the hypothetical future proposal may generalize the continuations to multi-shot.

Reply:
IMO, it is pretty unlikely that engine providers will support the extension to multi-shot continuations. 

There appear to be two primary issues:
1. No current 'language of interest' depends on multi-shot continuations.
2. Wasm is typically embedded in other environments, and those environments are dominated by languages like C/C++/Rust. It seems unlikely that multi-shot continuations can be made compatible with those implementation languages.